### PR TITLE
WIP: Changes to kernel cuda launch policies and procedure

### PR DESCRIPTION
**Pull request by MrBurmark**
Get max/recommended blocks depending on used num threads instead
of max threads.
Cuda occupancy calculator results cached to lower overhead.
Change occupancy launch to be more like fixed in that a max
num threads and num blocks may be specified or left
unspecified by using 0.
Kernel cuda dims fitting may now be done twice, first with
the recommended sizes and if that fails with the maximum sizes.

### CODE REVIEW

**Comment by ajkunen** [comment]
I'm taking it out for a spin

**Comment by MrBurmark** [comment]
@ajkunen @rhornung67, Can I get a review?

File: include/RAJA/policy/cuda/kernel/CudaKernel.hpp [hunk 1]
```diff
@@ -182,77 +220,126 @@ struct CudaLaunchHelper;
...
+    } else {
 
-  using executor_t = internal::cuda_statement_list_executor_t<StmtList, Data>;
+      if (num_threads <= 0) {
+
+        //
+        // determine threads at runtime, unsure what use 1024
+        // this value may be invalid for kernels with high register pressure
+        //
+        recommended_threads = 1024;
```
Comment (MrBurmark) [inline include/RAJA/policy/cuda/kernel/CudaKernel.hpp]: It's set here and below in max_threads

Comment (rhornung67) [reply include/RAJA/policy/cuda/kernel/CudaKernel.hpp]: Yeah, I saw that after I wrote my comment. I should have resolved it. Thanks!

**Pull request merged**